### PR TITLE
Enables probe systemdunitproperty to work offline

### PR DESCRIPTION
**Pull request by WesleyCeraso**
This implementation allows the probe to retrieve in offline mode only
the property UnitFileState.

### CODE REVIEW

File: src/OVAL/probes/unix/linux/systemdunitproperty.c [hunk 1]
```diff
@@ -251,6 +259,44 @@ static int unit_callback(const char *unit, void *cbarg)
 	return 0;
 }
 
+void get_unit_file_state_in_offline_mode(const char *unit, struct unit_callback_vars *vars)
+{
```
Comment (jan-cerny) [inline src/OVAL/probes/unix/linux/systemdunitproperty.c]: variable `unit` is not used

File: src/OVAL/probes/unix/linux/systemdunitproperty.c [hunk 2]
```diff
@@ -263,38 +309,55 @@ int probe_main(probe_ctx *ctx, void *probe_arg)
...
+		// Allows the probe to retrieve property UnitFileState even if systemd is not accessible
+		SEXP_t *temp_unit_entity, *temp_property_entity;
+		const char *cstr_unit_entity = SEXP_string_cstr(temp_unit_entity = probe_ent_getval(unit_entity));
+		SEXP_free(temp_unit_entity);
+		const char *cstr_property_entity = SEXP_string_cstr(temp_property_entity = probe_ent_getval(property_entity));
+		SEXP_free(temp_property_entity);
+		
+		if(strcmp(cstr_property_entity, UNIT_FILE_STATE) != 0) {
+			dbus_error_free(&dbus_error);
+			SEXP_t *msg = probe_msg_creat(OVAL_MESSAGE_LEVEL_INFO, "DBus connection failed, could not identify systemd units.");
```
Comment (jan-cerny) [inline src/OVAL/probes/unix/linux/systemdunitproperty.c]: This talks about DBus connection, but we don't do DBus connection here

**Comment by jan-cerny** [comment]
We're looking for a volunteer to take this pull request, I think only minor changes are needed. Any help would be much appreciated :heart:

File: src/OVAL/probes/unix/linux/systemdunitproperty.c [hunk 3]
```diff
@@ -251,6 +259,44 @@ static int unit_callback(const char *unit, void *cbarg)
 	return 0;
 }
 
+void get_unit_file_state_in_offline_mode(const char *unit, struct unit_callback_vars *vars)
+{
+	char systemctl_call[] = "systemctl list-unit-files | grep -P '\\S+\\.service'";
```
Comment (jan-cerny) [inline src/OVAL/probes/unix/linux/systemdunitproperty.c]: This ***cannot** be here. We can't call external programs!

**Comment by matejak** [comment]
I can adopt this PR.

File: src/OVAL/probes/unix/linux/systemdunitproperty.c (hunk 3 above)
Comment (matejak) [reply src/OVAL/probes/unix/linux/systemdunitproperty.c]: Is replacing this by calls to the DBUS a viable alternative?

**Comment by jan-cerny** [comment]
That's great!

File: src/OVAL/probes/unix/linux/systemdunitproperty.c (hunk 3 above)
Comment (jan-cerny) [reply src/OVAL/probes/unix/linux/systemdunitproperty.c]: I think we  have to implement something filesystem based.

We already do something like that eg. in process probe, where we read form '/proc/' directory to get attributes of processes.

You can try to have a look how systemd implements `systemctl list-unit-files`.

From a quick look I found this:
https://github.com/systemd/systemd/blob/fbd0b64f44a7bae678137d67ad5da202f6b8d809/src/systemctl/systemctl.c

However I don't know if their code can be reused, that needs to be investigated.

**Comment by matejak** [comment]
Code in this PR can't be used for the task, because we can't call `systemctl`. The proper way is to do the same as `systemd`, preferably using `systemd` routines.

**Pull request closed without merging**